storage/pebble/register_pruner_run.go

```diff
@@ -0,0 +1,120 @@
+package pebble
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/cockroachdb/pebble"
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// DBPruner is responsible for pruning outdated register entries in a Pebble database.
+// It batches deletions to improve performance and reduce the load on the system.
+type DBPruner struct {
+	logger zerolog.Logger
+	// pruneThrottleDelay controls a small pause between batches of registers inspected and pruned
+	pruneThrottleDelay time.Duration
+	pruneHeight        uint64
+
+	db                   *pebble.DB
+	lastRegisterID       flow.RegisterID
+	keepFirstRelevantKey bool
+
+	totalKeysPruned int
+}
+
+func RegisterPrunerRun(db *pebble.DB, logger zerolog.Logger, pruneThrottleDelay time.Duration, pruneHeight uint64) *DBPruner {
+	return &DBPruner{
+		logger:             logger,
+		pruneThrottleDelay: pruneThrottleDelay,
+		pruneHeight:        pruneHeight,
+		db:                 db,
+		totalKeysPruned:    0,
+	}
+}
+
+// BatchDelete deletes the provided keys from the database in a single batch operation.
+// It resets the batch for reuse, deletes each key from the batch, and commits the batch to ensure
+// that the deletions are applied atomically.
+//
+// Parameters:
+//   - ctx: The context for managing the pruning throttle delay operation.
+//   - lookupKeys: A slice of keys to be deleted from the database.
+//
+// No errors are expected during normal operations.
+func (p *DBPruner) BatchDelete(ctx context.Context, lookupKeys [][]byte) error {
+	dbBatch := p.db.NewBatch()
+	defer func() {
+		if cerr := dbBatch.Close(); cerr != nil {
+			p.logger.Err(cerr).Msg("error while closing the db batch")
+		}
+	}()
+
+	for _, key := range lookupKeys {
+		if err := dbBatch.Delete(key, nil); err != nil {
+			keyHeight, registerID, _ := lookupKeyToRegisterID(key)
+			return fmt.Errorf("failed to delete lookupKey (height: %d, registerID: %v): %w", keyHeight, registerID, err)
+		}
+	}
+
+	if err := dbBatch.Commit(pebble.Sync); err != nil {
+		return fmt.Errorf("failed to commit batch: %w", err)
+	}
+
+	p.totalKeysPruned += len(lookupKeys)
+
+	return nil
+}
+
+// CanPruneKey checks if a key can be pruned based on its height and the last processed register ID.
+// It ensures that only the first relevant key (the earliest entry) for each register ID is kept.
+//
+// Parameters:
+//   - key: The key to check for pruning eligibility.
+//
+// No errors are expected during normal operations.
+func (p *DBPruner) CanPruneKey(key []byte) (bool, error) {
+	keyHeight, registerID, err := lookupKeyToRegisterID(key)
+	if err != nil {
+		return false, fmt.Errorf("malformed lookup key %v: %w", key, err)
+	}
+
+	// If height is greater than prune height, the key cannot be pruned.
+	if keyHeight > p.pruneHeight {
+		return false, nil
```

Comment:
Should we check `if p.lastRegisterID != registerID` and update the state first? Would `lastRegisterID` get stale?

Please add comments to `lastRegisterID` and `keepFrstRelevantKey`

Reply:
I think this is OK to keep this first since we should ignore any row from a height that is not yet pruned. we only need to track `keepFirstRelevantKey` for heights <= pruneHeight.

+1 on adding comments. this logic is complex enough as it is, so let's make it easier for ourselves a year from now by adding extensive comments to anything that's non-obvious